File: typescript/src/clients/knowledge.ts
// =============================================================================
// Task API SDK - Knowledge Base Client (proxied to v2 API)
// =============================================================================

import { BaseClient } from "../client-base.js";
import type {
  AddDocumentRequest,
  AddDocumentResponse,
  CreateKnowledgeBaseRequest,
  CreateKnowledgeBaseResponse,
  DeleteDocumentsRequest,
  DeleteDocumentsResponse,
  GetDocumentResponse,
  GetKnowledgeBaseResponse,
  GetUploadUrlResponse,
  KnowledgeBaseInfo,
  ListFilesResponse,
  PaginatedResponse,
  QueryKnowledgeBaseRequest,
  QueryKnowledgeBaseResponse,
  RegisterFileUploadRequest,
  RegisterFileUploadResponse,
  RequestOptions,
  UploadFileResponse,
} from "../types.js";

/** Base path for the v2 knowledge base API. */
const KB_PATH = "/v2/knowledge";

/**
 * KnowledgeClient provides access to the Opper Knowledge Base API (v2).
 *
 * Knowledge bases allow you to store, index, and query documents using
 * vector embeddings for semantic search.
 */
export class KnowledgeClient extends BaseClient {
  // ---------------------------------------------------------------------------
  // Knowledge Base CRUD
  // ---------------------------------------------------------------------------

  /**
   * Create a new knowledge base.
   *
   * @param body - Name and optional embedding model configuration.
   * @param options - Optional request options.
   * @returns The created knowledge base.
   */
  async create(
    body: CreateKnowledgeBaseRequest,
    options?: RequestOptions,
  ): Promise<CreateKnowledgeBaseResponse> {
    return this.post<CreateKnowledgeBaseResponse>(KB_PATH, body, options);
  }

  /**
   * List all knowledge bases with pagination.
   *
   * @param params - Optional offset and limit for pagination.
   * @param options - Optional request options.
   * @returns Paginated list of knowledge bases.
   */
  async list(
    params?: { offset?: number; limit?: number },
    options?: RequestOptions,
  ): Promise<PaginatedResponse<KnowledgeBaseInfo>> {
    return this.get<PaginatedResponse<KnowledgeBaseInfo>>(KB_PATH, params, options);
  }

  /**
   * Get a knowledge base by ID.
   *
   * @param id - The knowledge base UUID.
   * @param options - Optional request options.
   * @returns Knowledge base details including document count.
   */
  async getById(
    id: string,
    options?: RequestOptions,
  ): Promise<GetKnowledgeBaseResponse> {
    return this.get<GetKnowledgeBaseResponse>(`${KB_PATH}/${id}`, undefined, options);
  }

  /**
   * Get a knowledge base by name.
   *
   * @param name - The knowledge base name.
   * @param options - Optional request options.
   * @returns Knowledge base details including document count.
   */
  async getByName(
    name: string,
    options?: RequestOptions,
  ): Promise<GetKnowledgeBaseResponse> {
    return this.get<GetKnowledgeBaseResponse>(
      `${KB_PATH}/by-name/${encodeURIComponent(name)}`,
      undefined,
      options,
    );
  }

  /**
   * Delete a knowledge base and all its contents.
   *
   * @param id - The knowledge base UUID.
   * @param options - Optional request options.
   */
  async deleteKnowledgeBase(id: string, options?: RequestOptions): Promise<void> {
    return this.delete(`${KB_PATH}/${id}`, options);
  }

  // ---------------------------------------------------------------------------
  // Document Operations
  // ---------------------------------------------------------------------------

  /**
   * Add a text document to a knowledge base.
   *
   * The document will be chunked and embedded automatically.
   *
   * @param knowledgeBaseId - The knowledge base UUID.
   * @param body - Document content, optional key, metadata, and chunking configuration.
   * @param options - Optional request options.
   * @returns The added document metadata.
   */
  async add(
    knowledgeBaseId: string,
    body: AddDocumentRequest,
    options?: RequestOptions,
  ): Promise<AddDocumentResponse> {
    return this.post<AddDocumentResponse>(`${KB_PATH}/${knowledgeBaseId}/add`, body, options);
  }

  /**
   * Query a knowledge base using semantic search.
   *
   * @param knowledgeBaseId - The knowledge base UUID.
   * @param body - Query string, filters, and result configuration.
   * @param options - Optional request options.
   * @returns Array of matching documents with relevance scores.
   */
  async query(
    knowledgeBaseId: string,
    body: QueryKnowledgeBaseRequest,
    options?: RequestOptions,
  ): Promise<QueryKnowledgeBaseResponse[]> {
    return this.post<QueryKnowledgeBaseResponse[]>(
      `${KB_PATH}/${knowledgeBaseId}/query`,
      body,
      options,
    );
  }

  /**
   * Get a document by its key.
   *
   * @param knowledgeBaseId - The knowledge base UUID.
   * @param documentKey - The document key.
   * @param options - Optional request options.
   * @returns The document with its segments and metadata.
   */
  async getDocument(
    knowledgeBaseId: string,
    documentKey: string,
    options?: RequestOptions,
  ): Promise<GetDocumentResponse> {
    return this.get<GetDocumentResponse>(
      `${KB_PATH}/${knowledgeBaseId}/documents/${encodeURIComponent(documentKey)}`,
      undefined,
      options,
    );
  }

  /**
   * Delete documents from a knowledge base using optional filters.
   * If no filters are provided, all documents are deleted.
   *
   * @param knowledgeBaseId - The knowledge base UUID.
   * @param body - Optional filters to select documents for deletion.
   * @param options - Optional request options.
   * @returns The count of deleted documents.
   */
  async deleteDocuments(
    knowledgeBaseId: string,
    body?: DeleteDocumentsRequest,
    options?: RequestOptions,
  ): Promise<DeleteDocumentsResponse> {
    return this.request<DeleteDocumentsResponse>("DELETE", `${KB_PATH}/${knowledgeBaseId}/query`, {
      ...options,
      body,
    });
  }

  // ---------------------------------------------------------------------------
  // File Operations
  // ---------------------------------------------------------------------------

  /**
   * Upload a file directly to a knowledge base.
   *
   * The file will be processed, chunked, and indexed automatically.
   * Supports PDF, TXT, DOCX, MD, HTML, RTF, and more (max 50MB).
   *
   * @param knowledgeBaseId - The knowledge base UUID.
   * @param file - The file to upload (Blob, File, or Buffer).
   * @param params - Optional chunking configuration and metadata.
   * @param options - Optional request options.
   * @returns The uploaded file metadata.
   */
  async uploadFile(
    knowledgeBaseId: string,
    file: Blob,
    params?: {
      filename?: string;
      chunkSize?: number;
      chunkOverlap?: number;
      metadata?: Record<string, unknown>;
    },
    options?: RequestOptions,
  ): Promise<UploadFileResponse> {
    const formData = new FormData();
    formData.append("file", file, params?.filename);
    if (params?.chunkSize != null) {
      formData.append("text_processing.chunk_size", String(params.chunkSize));
    }
    if (params?.chunkOverlap != null) {
      formData.append("text_processing.chunk_overlap", String(params.chunkOverlap));
    }
    if (params?.metadata != null) {
      formData.append("metadata", JSON.stringify(params.metadata));
    }

    const url = `${this.baseUrl}${KB_PATH}/${knowledgeBaseId}/upload`;
    const headers = {
      Authorization: `Bearer ${this.apiKey}`,
      ...this.defaultHeaders,
      ...options?.headers,
    };
    // Remove Content-Type so fetch sets the multipart boundary automatically
    delete (headers as Record<string, string>)["Content-Type"];

    const response = await fetch(url, {
      method: "POST",
      headers,
      body: formData,
      signal: options?.signal,
    });

    if (!response.ok) {
      let body: unknown;
      try {
        body = await response.json();
      } catch {
        body = await response.text().catch(() => undefined);
      }
      const { ApiError } = await import("../types.js");
      throw new ApiError(response.status, response.statusText, body);
    }

    return (await response.json()) as UploadFileResponse;
  }

  /**
   * Get a presigned URL for uploading a file to S3.
   *
   * This is step 1 of the 3-step upload process:
   * 1. Get upload URL → 2. Upload to S3 → 3. Register file
   *
   * @param knowledgeBaseId - The knowledge base UUID.
   * @param filename - The name of the file to upload.
   * @param options - Optional request options.
   * @returns Presigned URL, form fields, and file ID.
   */
  async getUploadUrl(
    knowledgeBaseId: string,
    filename: string,
    options?: RequestOptions,
  ): Promise<GetUploadUrlResponse> {
    return this.get<GetUploadUrlResponse>(
      `${KB_PATH}/${knowledgeBaseId}/upload_url`,
      { filename },
      options,
    );
  }

  /**
   * Register a file after uploading it to S3.
   *
   * This is step 3 of the 3-step upload process.
   *
   * @param knowledgeBaseId - The knowledge base UUID.
   * @param body - File registration details.
   * @param options - Optional request options.
   * @returns The registered file metadata.
   */
  async registerFileUpload(
    knowledgeBaseId: string,
    body: RegisterFileUploadRequest,
    options?: RequestOptions,
  ): Promise<RegisterFileUploadResponse> {
    return this.post<RegisterFileUploadResponse>(
      `${KB_PATH}/${knowledgeBaseId}/register_file`,
      body,
      options,
    );
  }

  /**
   * List files in a knowledge base with pagination.
   *
   * @param knowledgeBaseId - The knowledge base UUID.
   * @param params - Optional offset and limit for pagination.
   * @param options - Optional request options.
   * @returns Paginated list of files.
   */
  async listFiles(
    knowledgeBaseId: string,
    params?: { offset?: number; limit?: number },
    options?: RequestOptions,
  ): Promise<PaginatedResponse<ListFilesResponse>> {
    return this.get<PaginatedResponse<ListFilesResponse>>(
      `${KB_PATH}/${knowledgeBaseId}/files`,
      params,
      options,
    );
  }

  /**
   * Get a presigned download URL for a file.
   *
   * @param knowledgeBaseId - The knowledge base UUID.
   * @param fileId - The file UUID.
   * @param options - Optional request options.
   * @returns Object containing the download URL.
   */
  async getFileDownloadUrl(
    knowledgeBaseId: string,
    fileId: string,
    options?: RequestOptions,
  ): Promise<{ url: string }> {
    return this.get<{ url: string }>(
      `${KB_PATH}/${knowledgeBaseId}/files/${fileId}/download_url`,
      undefined,
      options,
    );
  }

  /**
   * Delete a file from a knowledge base.
   *
   * @param knowledgeBaseId - The knowledge base UUID.
   * @param fileId - The file UUID.
   * @param options - Optional request options.
   */
  async deleteFile(
    knowledgeBaseId: string,
    fileId: string,
    options?: RequestOptions,
  ): Promise<void> {
    return this.delete(`${KB_PATH}/${knowledgeBaseId}/files/${fileId}`, options);
  }
}

File: python/src/opperai/clients/knowledge.py
"""Opper SDK — Knowledge Base Client (v2 API)."""

from __future__ import annotations

from typing import Any
from urllib.parse import quote

from .._base_client import BaseClient
from ..types import (
    AddDocumentResponse,
    CreateKnowledgeBaseResponse,
    DeleteDocumentsResponse,
    FileInfo,
    GetDocumentResponse,
    GetKnowledgeBaseResponse,
    GetUploadUrlResponse,
    KnowledgeBaseInfo,
    QueryResult,
    RegisterFileUploadResponse,
    RequestOptions,
    UploadFileResponse,
)

KB_PATH = "/v2/knowledge"


class KnowledgeClient:
    """Client for the Knowledge Base API (v2)."""

    def __init__(self, client: BaseClient) -> None:
        self._client = client

    # --- Knowledge Base CRUD --------------------------------------------------

    def create(
        self,
        *,
        name: str,
        embedding_model: str | None = None,
        options: RequestOptions | None = None,
    ) -> CreateKnowledgeBaseResponse:
        body: dict[str, Any] = {"name": name}
        if embedding_model is not None:
            body["embedding_model"] = embedding_model
        data = self._client._post(KB_PATH, body, options=options)
        return CreateKnowledgeBaseResponse(**data)

    async def create_async(
        self,
        *,
        name: str,
        embedding_model: str | None = None,
        options: RequestOptions | None = None,
    ) -> CreateKnowledgeBaseResponse:
        body: dict[str, Any] = {"name": name}
        if embedding_model is not None:
            body["embedding_model"] = embedding_model
        data = await self._client._post_async(KB_PATH, body, options=options)
        return CreateKnowledgeBaseResponse(**data)

    def list(
        self,
        *,
        offset: int | None = None,
        limit: int | None = None,
        options: RequestOptions | None = None,
    ) -> list[KnowledgeBaseInfo]:
        query: dict[str, Any] = {}
        if offset is not None:
            query["offset"] = offset
        if limit is not None:
            query["limit"] = limit
        data = self._client._get(KB_PATH, query=query, options=options)
        return [KnowledgeBaseInfo(**kb) for kb in (data or {}).get("data", [])]

    async def list_async(
        self,
        *,
        offset: int | None = None,
        limit: int | None = None,
        options: RequestOptions | None = None,
    ) -> list[KnowledgeBaseInfo]:
        query: dict[str, Any] = {}
        if offset is not None:
            query["offset"] = offset
        if limit is not None:
            query["limit"] = limit
        data = await self._client._get_async(KB_PATH, query=query, options=options)
        return [KnowledgeBaseInfo(**kb) for kb in (data or {}).get("data", [])]

    def get(self, id: str, *, options: RequestOptions | None = None) -> GetKnowledgeBaseResponse:
        data = self._client._get(f"{KB_PATH}/{quote(id, safe='')}", options=options)
        return GetKnowledgeBaseResponse(**data)

    async def get_async(self, id: str, *, options: RequestOptions | None = None) -> GetKnowledgeBaseResponse:
        data = await self._client._get_async(f"{KB_PATH}/{quote(id, safe='')}", options=options)
        return GetKnowledgeBaseResponse(**data)

    def get_by_name(self, name: str, *, options: RequestOptions | None = None) -> GetKnowledgeBaseResponse:
        data = self._client._get(f"{KB_PATH}/by-name/{quote(name, safe='')}", options=options)
        return GetKnowledgeBaseResponse(**data)

    async def get_by_name_async(self, name: str, *, options: RequestOptions | None = None) -> GetKnowledgeBaseResponse:
        data = await self._client._get_async(f"{KB_PATH}/by-name/{quote(name, safe='')}", options=options)
        return GetKnowledgeBaseResponse(**data)

    def delete(self, id: str, *, options: RequestOptions | None = None) -> None:
        self._client._delete(f"{KB_PATH}/{quote(id, safe='')}", options=options)

    async def delete_async(self, id: str, *, options: RequestOptions | None = None) -> None:
        await self._client._delete_async(f"{KB_PATH}/{quote(id, safe='')}", options=options)

    # --- Documents ------------------------------------------------------------

    def add(
        self,
        kb_id: str,
        *,
        content: str,
        key: str | None = None,
        metadata: dict[str, Any] | None = None,
        chunk_size: int | None = None,
        chunk_overlap: int | None = None,
        options: RequestOptions | None = None,
    ) -> AddDocumentResponse:
        body: dict[str, Any] = {"content": content}
        if key is not None:
            body["key"] = key
        if metadata is not None:
            body["metadata"] = metadata
        config: dict[str, Any] = {}
        if chunk_size is not None:
            config["chunk_size"] = chunk_size
        if chunk_overlap is not None:
            config["chunk_overlap"] = chunk_overlap
        if config:
            body["configuration"] = config
        data = self._client._post(f"{KB_PATH}/{quote(kb_id, safe='')}/documents", body, options=options)
        return AddDocumentResponse(**data)

    async def add_async(
        self,
        kb_id: str,
        *,
        content: str,
        key: str | None = None,
        metadata: dict[str, Any] | None = None,
        chunk_size: int | None = None,
        chunk_overlap: int | None = None,
        options: RequestOptions | None = None,
    ) -> AddDocumentResponse:
        body: dict[str, Any] = {"content": content}
        if key is not None:
            body["key"] = key
        if metadata is not None:
            body["metadata"] = metadata
        config: dict[str, Any] = {}
        if chunk_size is not None:
            config["chunk_size"] = chunk_size
        if chunk_overlap is not None:
            config["chunk_overlap"] = chunk_overlap
        if config:
            body["configuration"] = config
        data = await self._client._post_async(f"{KB_PATH}/{quote(kb_id, safe='')}/documents", body, options=options)
        return AddDocumentResponse(**data)

    def query(
        self,
        kb_id: str,
        *,
        query: str,
        prefilter_limit: int | None = None,
        top_k: int | None = None,
        filters: list[dict[str, Any]] | None = None,
        rerank: bool | None = None,
        parent_span_id: str | None = None,
        options: RequestOptions | None = None,
    ) -> list[QueryResult]:
        body: dict[str, Any] = {"query": query}
        if prefilter_limit is not None:
            body["prefilter_limit"] = prefilter_limit
        if top_k is not None:
            body["top_k"] = top_k
        if filters is not None:
            body["filters"] = filters
        if rerank is not None:
            body["rerank"] = rerank
        if parent_span_id is not None:
            body["parent_span_id"] = parent_span_id
        data = self._client._post(f"{KB_PATH}/{quote(kb_id, safe='')}/query", body, options=options)
        if isinstance(data, list):
            return [QueryResult(**r) for r in data]
        return []

    async def query_async(
        self,
        kb_id: str,
        *,
        query: str,
        prefilter_limit: int | None = None,
        top_k: int | None = None,
        filters: list[dict[str, Any]] | None = None,
        rerank: bool | None = None,
        parent_span_id: str | None = None,
        options: RequestOptions | None = None,
    ) -> list[QueryResult]:
        body: dict[str, Any] = {"query": query}
        if prefilter_limit is not None:
            body["prefilter_limit"] = prefilter_limit
        if top_k is not None:
            body["top_k"] = top_k
        if filters is not None:
            body["filters"] = filters
        if rerank is not None:
            body["rerank"] = rerank
        if parent_span_id is not None:
            body["parent_span_id"] = parent_span_id
        data = await self._client._post_async(f"{KB_PATH}/{quote(kb_id, safe='')}/query", body, options=options)
        if isinstance(data, list):
            return [QueryResult(**r) for r in data]
        return []

    def get_document(self, kb_id: str, key: str, *, options: RequestOptions | None = None) -> GetDocumentResponse:
        data = self._client._get(f"{KB_PATH}/{quote(kb_id, safe='')}/documents/{quote(key, safe='')}", options=options)
        return GetDocumentResponse(**data)

    async def get_document_async(
        self, kb_id: str, key: str, *, options: RequestOptions | None = None
    ) -> GetDocumentResponse:
        data = await self._client._get_async(
            f"{KB_PATH}/{quote(kb_id, safe='')}/documents/{quote(key, safe='')}", options=options
        )
        return GetDocumentResponse(**data)

    def delete_documents(
        self,
        kb_id: str,
        *,
        filters: list[dict[str, Any]] | None = None,
        options: RequestOptions | None = None,
    ) -> DeleteDocumentsResponse:
        body: dict[str, Any] = {}
        if filters is not None:
            body["filters"] = filters
        data = self._client._delete(f"{KB_PATH}/{quote(kb_id, safe='')}/documents", body=body, options=options)
        return DeleteDocumentsResponse(**(data or {}))

    async def delete_documents_async(
        self,
        kb_id: str,
        *,
        filters: list[dict[str, Any]] | None = None,
        options: RequestOptions | None = None,
    ) -> DeleteDocumentsResponse:
        body: dict[str, Any] = {}
        if filters is not None:
            body["filters"] = filters
        data = await self._client._delete_async(
            f"{KB_PATH}/{quote(kb_id, safe='')}/documents", body=body, options=options
        )
        return DeleteDocumentsResponse(**(data or {}))

    # --- File Operations ------------------------------------------------------

    def upload_file(
        self,
        kb_id: str,
        file: Any,
        *,
        filename: str | None = None,
        chunk_size: int | None = None,
        chunk_overlap: int | None = None,
        metadata: dict[str, Any] | None = None,
        options: RequestOptions | None = None,
    ) -> UploadFileResponse:
        fields: dict[str, str] = {}
        if chunk_size is not None:
            fields["chunkSize"] = str(chunk_size)
        if chunk_overlap is not None:
            fields["chunkOverlap"] = str(chunk_overlap)
        if metadata is not None:
            import json

            fields["metadata"] = json.dumps(metadata)
        data = self._client._upload(
            f"{KB_PATH}/{quote(kb_id, safe='')}/files/upload",
            file,
            filename=filename,
            fields=fields,
            options=options,
        )
        return UploadFileResponse(**data)

    async def upload_file_async(
        self,
        kb_id: str,
        file: Any,
        *,
        filename: str | None = None,
        chunk_size: int | None = None,
        chunk_overlap: int | None = None,
        metadata: dict[str, Any] | None = None,
        options: RequestOptions | None = None,
    ) -> UploadFileResponse:
        fields: dict[str, str] = {}
        if chunk_size is not None:
            fields["chunkSize"] = str(chunk_size)
        if chunk_overlap is not None:
            fields["chunkOverlap"] = str(chunk_overlap)
        if metadata is not None:
            import json

            fields["metadata"] = json.dumps(metadata)
        data = await self._client._upload_async(
            f"{KB_PATH}/{quote(kb_id, safe='')}/files/upload",
            file,
            filename=filename,
            fields=fields,
            options=options,
        )
        return UploadFileResponse(**data)

    def get_upload_url(
        self, kb_id: str, *, filename: str, options: RequestOptions | None = None
    ) -> GetUploadUrlResponse:
        data = self._client._post(
            f"{KB_PATH}/{quote(kb_id, safe='')}/files/upload-url",
            {"filename": filename},
            options=options,
        )
        return GetUploadUrlResponse(**data)

    async def get_upload_url_async(
        self, kb_id: str, *, filename: str, options: RequestOptions | None = None
    ) -> GetUploadUrlResponse:
        data = await self._client._post_async(
            f"{KB_PATH}/{quote(kb_id, safe='')}/files/upload-url",
            {"filename": filename},
            options=options,
        )
        return GetUploadUrlResponse(**data)

    def register_file(
        self,
        kb_id: str,
        *,
        filename: str,
        file_id: str,
        content_type: str,
        chunk_size: int | None = None,
        chunk_overlap: int | None = None,
        metadata: dict[str, Any] | None = None,
        options: RequestOptions | None = None,
    ) -> RegisterFileUploadResponse:
        body: dict[str, Any] = {"filename": filename, "file_id": file_id, "content_type": content_type}
        config: dict[str, Any] = {}
        if chunk_size is not None:
            config["chunk_size"] = chunk_size
        if chunk_overlap is not None:
            config["chunk_overlap"] = chunk_overlap
        if config:
            body["configuration"] = config
        if metadata is not None:
            body["metadata"] = metadata
        data = self._client._post(f"{KB_PATH}/{quote(kb_id, safe='')}/files/register", body, options=options)
        return RegisterFileUploadResponse(**data)

    async def register_file_async(
        self,
        kb_id: str,
        *,
        filename: str,
        file_id: str,
        content_type: str,
        chunk_size: int | None = None,
        chunk_overlap: int | None = None,
        metadata: dict[str, Any] | None = None,
        options: RequestOptions | None = None,
    ) -> RegisterFileUploadResponse:
        body: dict[str, Any] = {"filename": filename, "file_id": file_id, "content_type": content_type}
        config: dict[str, Any] = {}
        if chunk_size is not None:
            config["chunk_size"] = chunk_size
        if chunk_overlap is not None:
            config["chunk_overlap"] = chunk_overlap
        if config:
            body["configuration"] = config
        if metadata is not None:
            body["metadata"] = metadata
        data = await self._client._post_async(
            f"{KB_PATH}/{quote(kb_id, safe='')}/files/register", body, options=options
        )
        return RegisterFileUploadResponse(**data)

    def list_files(
        self,
        kb_id: str,
        *,
        offset: int | None = None,
        limit: int | None = None,
        options: RequestOptions | None = None,
    ) -> list[FileInfo]:
        query: dict[str, Any] = {}
        if offset is not None:
            query["offset"] = offset
        if limit is not None:
            query["limit"] = limit
        data = self._client._get(f"{KB_PATH}/{quote(kb_id, safe='')}/files", query=query, options=options)
        return [FileInfo(**f) for f in (data or {}).get("data", [])]

    async def list_files_async(
        self,
        kb_id: str,
        *,
        offset: int | None = None,
        limit: int | None = None,
        options: RequestOptions | None = None,
    ) -> list[FileInfo]:
        query: dict[str, Any] = {}
        if offset is not None:
            query["offset"] = offset
        if limit is not None:
            query["limit"] = limit
        data = await self._client._get_async(f"{KB_PATH}/{quote(kb_id, safe='')}/files", query=query, options=options)
        return [FileInfo(**f) for f in (data or {}).get("data", [])]

    def get_download_url(self, kb_id: str, file_id: str, *, options: RequestOptions | None = None) -> str:
        data = self._client._get(
            f"{KB_PATH}/{quote(kb_id, safe='')}/files/{quote(file_id, safe='')}/download-url",
            options=options,
        )
        return (data or {}).get("url", "")

    async def get_download_url_async(self, kb_id: str, file_id: str, *, options: RequestOptions | None = None) -> str:
        data = await self._client._get_async(
            f"{KB_PATH}/{quote(kb_id, safe='')}/files/{quote(file_id, safe='')}/download-url",
            options=options,
        )
        return (data or {}).get("url", "")

    def delete_file(self, kb_id: str, file_id: str, *, options: RequestOptions | None = None) -> None:
        self._client._delete(f"{KB_PATH}/{quote(kb_id, safe='')}/files/{quote(file_id, safe='')}", options=options)

    async def delete_file_async(self, kb_id: str, file_id: str, *, options: RequestOptions | None = None) -> None:
        await self._client._delete_async(
            f"{KB_PATH}/{quote(kb_id, safe='')}/files/{quote(file_id, safe='')}", options=options
        )
